subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java

```diff
@@ -189,25 +190,29 @@ private static boolean missesDependency(Node producer, Node consumer) {
         // Do a breadth first search for any dependency
         Deque<Node> queue = new ArrayDeque<>();
         Set<Node> seenNodes = new HashSet<>();
-        consumer.getHardSuccessors().forEach(successor -> {
-            if (seenNodes.add(successor)) {
-                queue.add(successor);
-            }
-        });
+        addHardSuccessorTasksToQueue(consumer, seenNodes, queue);
         while (!queue.isEmpty()) {
             Node dependency = queue.removeFirst();
             if (dependency == producer) {
                 return false;
             }
-            dependency.getHardSuccessors().forEach(node -> {
-                if (seenNodes.add(node)) {
-                    queue.add(node);
-                }
-            });
+            addHardSuccessorTasksToQueue(dependency, seenNodes, queue);
         }
         return true;
     }
 
+    private static void addHardSuccessorTasksToQueue(Node node, Set<Node> seenNodes, Queue<Node> queue) {
+        node.getHardSuccessors().forEach(successor -> {
+            if (successor instanceof LocalTaskNode) {
+                if (seenNodes.add(successor)) {
+                    queue.add(successor);
+                }
+            } else {
+                addHardSuccessorTasksToQueue(successor, seenNodes, queue);
+            }
+        });
+    }
```

Comment:
We've talked about introducing an `AddOnceQueue` implementation a couple months ago when working with snapshot hierarchy visiting IIRC. That would come in handy here, too.

Reply:
It would be nice to have some explanation in the code why we only handle `LocalTask` nodes here.

Reply:
I think we talked about the `AddOnceQueue` exactly for the code above.

I'll add a comment why we only consider `LocalTask` nodes here.

Reply:
How about this? https://github.com/gradle/gradle/blob/5cd9d4208ef0ca96c7ec5e7f53964c55e62d2c28/subprojects/core/src/main/java/org/gradle/execution/plan/LocalTaskNodeExecutor.java#L206-L208

Reply:
👍